src/cdk-experimental/testing/component-harness.ts

```diff
@@ -8,183 +8,267 @@
+   * component matching the given harness type under the root element of this `LocatorFactory`.
+   * When the resulting locator function is invoked, if multiple matching components are found, a
+   * harness for the first one is returned. If no components are found, an error is thrown.
+   * @param harnessType The type of harness to search for.
+   * @return An asynchronous locator function that searches components matching the given harness
+   *     type, and either returns a `ComponentHarness` for the component, or throws an error.
+   */
+  requiredLocator<T extends ComponentHarness>(harnessType: ComponentHarnessConstructor<T>):
+      AsyncFn<T>;
+
+  /**
+   * Creates an asynchronous locator function that can be used to search for elements with the given
+   * selector under the root element of this `LocatorFactory`. When the resulting locator function
+   * is invoked, if multiple matching elements are found, the first element is returned. If no
+   * elements are found, null is returned.
+   * @param selector The selector for the element that the locator function should search for.
+   * @return An asynchronous locator function that searches for elements with the given selector,
+   *     and either finds one or returns null.
+   */
+  optionalLocator(selector: string): AsyncFn<TestElement | null>;
+
+  /**
+   * Creates an asynchronous locator function that can be used to find a `ComponentHarness` for a
+   * component matching the given harness type under the root element of this `LocatorFactory`.
+   * When the resulting locator function is invoked, if multiple matching components are found, a
+   * harness for the first one is returned. If no components are found, null is returned.
+   * @param harnessType The type of harness to search for.
+   * @return An asynchronous locator function that searches components matching the given harness
+   *     type, and either returns a `ComponentHarness` for the component, or null if none is found.
+   */
+  optionalLocator<T extends ComponentHarness>(harnessType: ComponentHarnessConstructor<T>):
+      AsyncFn<T | null>;
 
   /**
-   * Generate a function to find the first matched test element by CSS
-   * selector.
-   * @param selector The CSS selector of the test element.
+   * Creates an asynchronous locator function that can be used to search for a list of elements with
+   * the given selector under the root element of this `LocatorFactory`. When the resulting locator
+   * function is invoked, a list of matching elements is returned.
+   * @param selector The selector for the element that the locator function should search for.
+   * @return An asynchronous locator function that searches for elements with the given selector,
+   *     and either finds one or throws an error
    */
-  protected find(selector: string): () => Promise<TestElement>;
+  allLocator(selector: string): AsyncFn<TestElement[]>;
```

Comment:
`allLocator` feels slightly odd to me. Thinking of alternatives along the lines of `multiLocator` or `locatorForAll`

Reply:
`locatorFor{Required,Optional,All}` was actually one of my alternative suggestions. It's slightly longer, but I think it reads better.